tools/checked-c-convert/ProgramInfo.cpp

```diff
@@ -31,6 +31,163 @@ bool ProgramInfo::checkStructuralEquality(uint32_t V, uint32_t U) {
+        
+        std::vector<std::set<uint32_t> > &pVars1 = (*I)->getParams();
+        std::vector<std::set<uint32_t> > &pVars2 = (*J)->getParams();
+        assert(pVars1.size() == pVars2.size());
+
+        for (std::vector<std::set<uint32_t> >::iterator V1 = pVars1.begin(),
+          V2 = pVars2.begin();
+          V1 != pVars1.end() && V2 != pVars2.end();
+          ++V1, ++V2)
+        {
+          std::set<uint32_t> pv1 = *V1;
+          std::set<uint32_t> pv2 = *V2;
+          assert(pv1.size() == pv2.size());
+
+          for (std::set<uint32_t>::iterator V1 = pv1.begin(), V2 = pv2.begin();
+            V1 != pv1.end() && V2 != pv2.end(); ++V1, ++V2)
+            CS.addConstraint(CS.createEq(
+              CS.getOrCreateVar(*V1), CS.getOrCreateVar(*V2)));
+        }
+      }
+    }
+
+    // Do the same as above, but in a simpler case where we only need to 
+    // constrain according to the type of the global variable. 
+    for (std::set<GlobalVariableSymbol*>::iterator I = vars.begin();
+      I != vars.end(); ++I) {
+
+    }
+  }
+
+  // For every global function that is an unresolved external, constrain 
+  // its parameter types to be wild.
+  for (const auto &UnkSymbol : ExternFunctions) {
+
+    std::map<std::string, std::set<GlobalSymbol*> >::iterator I = 
+      GlobalSymbols.find(UnkSymbol);
+    assert(I != GlobalSymbols.end());
+    std::set<GlobalSymbol*> Gs = (*I).second;
+
+    for (const auto &G : Gs) {
+      if (GlobalFunctionSymbol *GFS = dyn_cast<GlobalFunctionSymbol>(G)) {
+        for (const auto &V : GFS->getReturns())
+          CS.addConstraint(CS.createEq(CS.getOrCreateVar(V), CS.getWild()));
+
+        for (const auto &U : GFS->getParams())
+          for(const auto &V : U)
+            CS.addConstraint(CS.createEq(CS.getOrCreateVar(V), CS.getWild()));
+      }
+    }
+  }
+
+  return true;
+}
+
+void ProgramInfo::seeFunctionDecl(FunctionDecl *F, ASTContext *C) {
+  if (!F->isGlobal())
+    return;
+
+  // Look up the constraint variables for the return type and parameter 
+  // declarations of this function, if any.
+  std::string fn = F->getNameAsString();
+  std::set<uint32_t> returnVars;
+  std::vector<std::set<uint32_t> > parameterVars(F->getNumParams());
+  PersistentSourceLoc PLoc =
+    PersistentSourceLoc::mkPSL(F->getLocation(), *C);
+  int i = 0;
+
+  getVariable(F, returnVars, C);
+  for (auto &I : F->params())
+    getVariable(I, parameterVars[i++], C);
+
+  assert(PLoc.valid());
+  GlobalFunctionSymbol *GF = 
+    new GlobalFunctionSymbol(fn, PLoc, parameterVars, returnVars);
+
+  if (F->isThisDeclarationADefinition() && F->hasBody()) {
+    // Check if this function has a body and if it appears in our set of 
+    // functions with no body. If it does, remove it. 
+    std::set<std::string>::iterator I = ExternFunctions.find(fn);
+    if (I != ExternFunctions.end())
+      ExternFunctions.erase(I);
+  } else {
+    // If this function has no body, add it to the set of functions that
```

Comment:
A duplicate declaration could occur after the function has been defined.   You need some other way to check for that.   I think the code:

extern void f(void) {}

extern void f(void);

is valid.